### ISSUE TITLE
Overriding methods with defs

### ISSUE THREAD

**Issue by apblack** [gracelang/language#23]
We decided a while back that we would not permit the overriding of a method with a def.  The motivation was that the initialization code in the superobject might request that method. If it is overridden with a def, then the variable will not be initialized when accessed by the superobject's initialization.  In this case the programmer will see a runtime Exception due to accessing the uninitialized variable.

This is a real problem, but one that is is part and parcel of the semantics that allows overriding of self in initialization.  The problem will still occur if the super-method is overridden by a sub-method that accesses a variable in the sub-object.  In other words, the restriction on overriding a method with a def does not avoid the problem.

This restriction rules out very reasonable code.  For example, in the (pure) superobject enumerable that is inherited by different kinds of enumerable collection, the method size is defined as follows:

```grace
    method size {
        // override if size is known
        SizeUnknown.raise "size requested on {asDebugString}"
    }
```
It might alsi be defined as
```grace    
    method size { abstract }
```

It is quite natural in a concrete subobject to want to define

```grace
    var size is readable := 0

    method add(element:T) {
        size := size + 1
        // code to insert element into self
    }
```

This is presently illegal because of the restriction on overriding a **method** with a **def**.  Instead one is supposed to to write:

```grace
    var size' := 0
 
    method size { size' }

    method add(element:T) {
        size' := size' + 1
        // code to insert element into self
    }
```

Naturally, if `size` _were_ accessed during the initialization of the super-object, this would still raise the un-initialized variable exception, this time on `size'`. 

In short, this restriction forces programmers to write circumlocutions that are ugly, violating our principle that Grace code should be clear and natural.   Yes, it does not avoid the problem that it was intended to avoid.  Thus, I propose that we drop this restriction, having learned from experience.

We should replace it with a discussion of the dangers of accessing self attributes from initialization code — a section that we need to write anyway.

**Comment by kjx** [comment]
I'm happy with this. I'm adding it to the general discussion on traits, which I hope to work on...

**Comment by apblack** [comment]
This has nothing whatsoever to do with traits.

**Comment by kjx** [comment]
What I meant was - once we agree on a semantics for traits (and class inheritance) it will contain an answer to this question.

**Comment by apblack** [comment]
We have agreed on a semantics for traits — just not a syntax ;-)

This issue comes up with inheritance, but _not_ with traits, because traits can't contain defs.

**Comment by kjx** [comment]
> We have agreed on a semantics for traits — just not a syntax ;-)

actually I think it's the other way around :-).
We don't have all the corner cases tied down - like this one. 

> This issue comes up with inheritance, but not with traits, because traits can't contain defs

It says "overriding methods with defs. Traits can have methods, and classes that use traits can have defs! 

> in the (pure) superobject enumerable

is a trait!! it's a trait! 

(I'm trying to do the wadlerfest proofread: This is more fun)

**Comment by apblack** [comment]
> is a trait!! it's a trait!

Yes, it's a trait now.  It wasn't a trait when I raised this issue, because we didn't have traits then.

So you are right, and this issue does come up with objects that use traits.  My point should have been that it comes up with objects that don't use traits too, and the origin of the original method (trait or superobject) is irrelevant to the discussion.

**Comment by kjx** [comment]
still open: I propose we remove this restriction.

**Comment by apblack** [comment]
This is now strongly connected to #33.  If we make **def**s introduce a lexical binding, rather than a method, then it makes no sense to say that a def can override a method, or a method a def, since they are different kinds of fruit.

**Comment by kjx** [comment]
And thus to #47.  I hope to talk about this this afternoon

**Comment by kjx** [comment]
drafted - no restrictions - https://github.com/gracelang/language/blob/portland-james/spec.md#reuse
